### PR TITLE
[TWP][1/n] Efficient __and__ and __or__ operators for TimeWindowPartitionsSubsets

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This builds off of the work in https://github.com/dagster-io/dagster/pull/22014, and implements __and__ and __or__ in a similar style.

### CODE REVIEW

**Comment by OwenKephart** [comment]
* **#22120** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22119** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22118** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22118?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py [hunk 1]
```diff
@@ -2256,6 +2256,80 @@ def with_partitions_def(
     def __repr__(self) -> str:
         return f"TimeWindowPartitionsSubset({self.get_partition_key_ranges(self.partitions_def)})"
 
+    def __and__(self, other: "PartitionsSubset") -> "PartitionsSubset":
+        if self == other:
+            return self
+
+        empty_subset = self.empty_subset(self.partitions_def)
+        if other is empty_subset:
+            return other
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:2265]: `return empty_subset` right? (add test case?)

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py [hunk 2]
```diff
@@ -2256,6 +2256,80 @@ def with_partitions_def(
...
+        while self_window and other_window:
+            # find the intersection between the current two windows
+            start = max(self_window.start, other_window.start)
+            end = min(self_window.end, other_window.end)
+
+            # these windows intersect
+            if start < end:
+                result_windows.append(TimeWindow(start=start, end=end))
+
+            # advance the iterator with the earliest end time to find the next potential intersection
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:2289]: something similar here might simplify the sub logic - there may be one case there that isn't strictly needed

the case that i had to think carefully about in the sub PR was where one window might need to run again on the window that is spat out after the operation is applied - which i think is applicable to sub() but not and()

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py [hunk 3]
```diff
@@ -2256,6 +2256,80 @@ def with_partitions_def(
...
+            return super().__or__(other)
+
+        input_time_windows = sorted(
+            [*self.included_time_windows, *other.included_time_windows],
+            key=lambda tw: tw.start.timestamp(),
+        )
+        result_windows = [input_time_windows[0]]
+        for window in input_time_windows[1:]:
+            latest_window = result_windows[-1]
+            if window.start < latest_window.end:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:2319]: should this not be <=? What if you OR [2020-01-01, 2020-01-02] and [2020-01-02, 2020-01-03], they should be combined right?

**Review by gibsondan — CHANGES_REQUESTED** [review]
couple small things - the high-level plan for both here makes sense to me

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py (hunk 3 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:2319]: That's right -- I think I accidentally fixed that upstack, bringing that down here now.

**Review by gibsondan — APPROVED** [review]
addressed downstack

**Comment by OwenKephart** [comment]
### Merge activity

* **May 28, 4:19 PM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22118).

**Pull request merged**